lib/puppet/pops/types/enumeration.rb

```diff
@@ -1,33 +1,72 @@
 # The Enumeration class provides default Enumerable::Enumerator creation for Puppet Programming Language
 # runtime objects that supports the concept of enumeration.
 #
-class Puppet::Pops::Types::Enumeration
-  # Produces an Enumerable::Enumerator for Array, Hash, Integer, Integer Range, and String.
+module Puppet::Pops::Types
+class Enumeration
+
+  # Produces an `Enumerator` for `String`, `Integer`, `PIntegerType` or anything that respond to `each` by
+  # returning an `Enumerator`. The argument is returned in case it already is an `Enumerator`.
+  #
+  # @param caller [Function] The function that needs to coerce its argument to an enumeration
+  # @param o [Object] The object to produce an `Enumerator` for
+  # @return [Enumerator] The produced `Enumerator` or nil if it couldn't be produced
+  # @raise [ArgumentError] In case an `Enumerator` cannot be produced
+  # @api public
+  def self.asserted_enumerable(caller, o)
+    enum = self.enumerator(o)
+    raise ArgumentError, "#{caller.class.name}(): wrong argument type (#{o.class.name}; must be something enumerable." if enum.nil?
+    enum
+  end
+
+  # Produces an `Enumerator` for `String`, `Integer`, `PIntegerType` or anything that respond to `each` by
+  # returning an `Enumerator`. The argument is returned in case it already is an `Enumerator`.
   #
+  # @param o [Object] The object to produce an `Enumerator` for
+  # @return [Enumerator,nil] The produced `Enumerator` or `nil` if it couldn't be produced
+  # @api public
   def self.enumerator(o)
     @@singleton ||= new
     @@singleton.enumerator(o)
   end
 
-  # Produces an Enumerator for Array, Hash, Integer, Integer Range, and String.
+  # Produces an `Enumerator` for `String`, `Integer`, `Range`, `PIntegerType` or anything that respond to `each` by
+  # returning an `Enumerator`. The argument is returned in case it already is an `Enumerator` or an `Iterable`.
   #
+  # @param o [Object] The object to produce an `Enumerator` for
+  # @return [Enumerator,nil] The produced `Enumerator` or `nil` if it couldn't be produced
+  # @api public
   def enumerator(o)
     case o
     when String
-      x = o.chars
-      # Ruby 1.9.3 returns an Enumerator, and 2.0.0 an Array
-      x.is_a?(Array) ? x.each : x
-    when Integer
-      o.times
+      EnumerationIterable.new(PStringType::DEFAULT, o.each_char)
     when Array
-      o.each
-    when Hash
-      o.each
-    when Puppet::Pops::Types::PIntegerType
-      # Not enumerable if representing an infinite range
-      o.enumerable? ? o.each : nil
+      EnumerationIterable.new(TypeCalculator.singleton.infer_and_reduce_type(o), o.each)
```

Comment:
This is how the TypeCalculator will infer the element type of an Array, and hence, also how the element type of the Iterator is calculated. It could be improved of course (the infer_and_reduce method), but it's not in scope for this PR IMO.